### DISCUSSION TITLE
Inefficient dataloader query with many-to-many

### DISCUSSION THREAD

**Discussion by jukkahell** [mikro-orm/mikro-orm#6126] [General]
### Describe the bug

Let's say we have Video and Category entities with m:n relation.

With graphql we need use dataloaders. If client asks for videos with categories the dataloader with `video.categories.load()` ends up with this kind of find call:
```typescript
await orm.em.find(Category, { videos: { $in: ['videoId'] } }, { populate: ['videos'] });
```
which leads to this kind of sql query:
```sql
SELECT `c0`.*,
       `v1`.`id` AS `v1__id`,
       `v1`.`name` AS `v1__name`
FROM `category` AS `c0`
LEFT JOIN `video_categories` AS `v2` ON `c0`.`id` = `v2`.`category_id`
LEFT JOIN `video` AS `v1` ON `v2`.`video_id` = `v1`.`id`
LEFT JOIN `video_categories` AS `v3` ON `c0`.`id` = `v3`.`category_id`
WHERE `v3`.`video_id` IN (1)
```

where the third left join is not relevant and may lead up to thousands of identical rows and query time is many seconds. In my usecase, I have about 100 categories and about 5000 video_category relations and it takes ~18 seconds if I it ends up in the categories dataloader.

I wasn't able to figure out myself if there is anything that could be done so hope to get some help.

### Reproduction

```typescript
import { Collection, Entity, ManyToMany, MikroORM, PrimaryKey, Property } from '@mikro-orm/core';
import { SqliteDriver } from '../../packages/sqlite/src';

@Entity()
class Video {

  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @ManyToMany(() => Category)
  categories = new Collection<Category>(this);

}

@Entity()
class Category {

  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @ManyToMany(() => Video, video => video.categories)
  videos = new Collection<Video>(this);

}

describe('Dataloader m:n', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Video, Category],
      dbName: ':memory:',
      driver: SqliteDriver,
      debug: true,
      dataloader: true,
      allowGlobalContext: true,
    });
    await orm.getSchemaGenerator().refreshDatabase();
  });

  afterAll(() => orm.close(true));

  test('Load m:n', async () => {
    const video = orm.em.create(Video, { name: 'video1', categories: [{ name: 'category1' }] });
    await orm.em.flush();
    orm.em.clear();

    // Similar find query for comparison
    // await orm.em.find(Category, { videos: { $in: [video.id] } }, { populate: ['videos'] });
    // orm.em.clear();
    const video1 = await orm.em.findOneOrFail(Video, video.id);
    await video1.categories.load();
  });

});
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.13

### Node.js version

20.12.2

### Operating system

MacOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by darkbasic** [comment]
@B4nan this reminds me one of the issues we've discussed when v6 was in development, probably one of the following I guess:
https://github.com/mikro-orm/mikro-orm/issues/5019
https://github.com/mikro-orm/mikro-orm/pull/5031
https://github.com/mikro-orm/mikro-orm/issues/5014

**Comment by B4nan** [comment]
> where the third left join is not relevant and may lead up to thousands of identical rows and query time is many seconds.

It is relevant, by default, you populate full relations regardless of your where condition. One join branch is used to filter root entity, the other for population (select). You can control this behavior via `populateWhere` option, both globally and locally, the default is `all` which uses this approach, if you use `infer` instead, the join branches will be shared for the population. I don't think this has any relevance for the dataloader itself, or do you see it behaving differently without dataloaders enabled?

**Comment by jukkahell** [comment]
`infer` seems to work. Thank you for your help!
